File: RemoteBees-iOS/RemoteBeesService/Couchbase/DictionaryEncoder.swift
/***************************************************************************
 * This source file is part of the RemoteBees open source project.         *
 *                                                                         *
 * Licensed under the MIT License. See LICENSE for license information     *
 ***************************************************************************/

import Foundation

class DictionaryEncoder {
    
    static var jsonEncoder: JSONEncoder {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"

        let jsonEncoder = JSONEncoder()
        jsonEncoder.dateEncodingStrategy = .formatted(formatter)
        return jsonEncoder
    }

    func encode<T>(_ value: T) throws -> [String: Any] where T : Encodable {
        let data = try Self.jsonEncoder.encode(value)
        return try JSONSerialization.jsonObject(with: data, options: .allowFragments) as! [String: Any]
    }

}

File: RemoteBees-Android/RemoteBeesService/src/main/java/dev/beehive/remotebees/service/internal/couchbase/DictionaryEncoder.kt
/***************************************************************************
 * This source file is part of the RemoteBees open source project.         *
 *                                                                         *
 * Licensed under the MIT License. See LICENSE for license information     *
 ***************************************************************************/

package dev.beehive.remotebees.service.internal.couchbase

import com.inmotionsoftware.httpservice.coder.EncoderException
import com.inmotionsoftware.httpservice.coder.URIJsonAdapter
import com.inmotionsoftware.httpservice.coder.URLJsonAdapter
import com.inmotionsoftware.httpservice.coder.UUIDJsonAdapter
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.adapters.Rfc3339DateJsonAdapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import java.util.*

class DictionaryEncoder {

    private val moshi: Moshi

    init {
        val builder = Moshi.Builder()
        builder.add(KotlinJsonAdapterFactory())
            .add(Date::class.java, Rfc3339DateJsonAdapter().nullSafe())
            .add(UUIDJsonAdapter())
            .add(URLJsonAdapter())
            .add(URIJsonAdapter())
        this.moshi = builder.build()
    }

    @Throws(EncoderException::class)
    fun <T:Any> encode(value: T): Map<String, Any>? {
        val adapter: JsonAdapter<T> = this.moshi.adapter(value.javaClass)
        return adapter.toJsonValue(value) as? Map<String, Any>
    }

}
